enhancements/machine-api/converting-machine-api-to-cluster-api.md

```diff
@@ -0,0 +1,754 @@
+
+#### Story 4
+
+As a developer of Machine related APIs, I would like to be able to gather feedback on Machine conversion layers before forcing new APIs onto users,
+so that we are confident when we go to GA the new APIs, that we will not break our existing users.
+
+### Goals
+
+* Allow users to migrate Machines between MAPI/CAPI to test the new API
+* Allow users a choice (initially[^3]) of when they want to migrate
+* Eventually[^3] remove the Machine API controllers and leverage the migration layer to convert MAPI machines to CAPI machines
+* Allow the Cluster Autoscaler, MachineHealthCheck and ControlPlaneMachineSet to continue to work during a semi-migrated cluster state
+* Ensure that every MAPI resource is translated to a CAPI resource to allow users to test the new API
+* Allow access to newer features of CAPI that do not exist in MAPI
+
+[^3]: In the first iteration of this project, the goal is to allow users to migrate individualy resources.
+In later iterations, the plan is to remove the Machine API controllers. The linked goals will be executed in separate releases.
+
+### Non-Goals
+
+* Setting up or configuring any part of Cluster API, this is handled in a separate enhancement
+* Modifying any of the existing deployments of Machine API within clusters
+* Deleting or modifying any existing resources within clusters (until a user chooses to do so)
+* Force creation of new MAPI machines when new CAPI MachineSets are being created[^4]
+* Port any feature from CAPI into MAPI
+
+[^4]: We want to promote, over time, a reduction in the number of MAPI Machines, as such,
+any resources created directly in CAPI, will not be reflected into MAPI, unless they have an owner,
+and, that owner is already reflected into MAPI.
+
+## Proposal
+
+We will implement a two-way synchronisation controller that synchronises Machines and related resources between the Machine API and Cluster API equivalents.
+
+Using an annotation on the Machine API resource, users will be able to choose which API is authoritative.
+Any changes to the non-authoritative resource will be overwritten by the synchronisation controller to match the authoritative resource.
+
+When the API is non-authoritative, the controllers should ignore the resource,
+allowing the authoritative API’s controller to perform the reconciliation actions required.
+
+### Workflow Description
+
+In this workflow, the **cluster admin** is responsible for the infrastructure provisioning for the cluster.
+
+When the cluster admin wishes to migrate a MachineSet from Machine API to Cluster API, the following procedure is required:
+1. Identify the MachineSet to migrate to Cluster API
+1. Use the `oc edit` or a patch command to update the value of the `machines.openshift.io/authoritative-api` annotation to `cluster.x-k8s.io`
+1. The validating webhook verifies that the value of the annotation is valid and that the synchronised generation is up to date, and permits the request
```

Comment:
The advantage is that it's explicit for every resource you add conversion support for. If it's not in the API there's no conversion support.

The spec/status approach implies that there's a conversion controller involved. In this case, this would presumably be handled by the two-way sync controller.

The user (or some agent) updates the MAPI spec to indicate an intent to make CAPI authoritative. At this point, MAPI is still authoritative.

The sync controller waits until:
* The MAPI resource is in a stable state according to resource-specific requirements
* The corresponding CAPI resource is up to date

The sync controller updates the status to indicate that CAPI is authoritative.

Reply:
I think having a spec/status field makes sense for handling this information. It might be awkward that we're extending an API that we want to deprecate, but since it's in service of that deprecation and, hopefully, becoming the primary use case for it when the current MAPI controllers are removed, it's not awful.

I like Matt's point that it makes conversion support very explicit, too. I can see someone adding the annotation to resources we've chosen not to support and getting confused that nothing's happening, even if the annotation can be successfully written.